helix-term/src/commands.rs

```diff
@@ -3613,6 +3613,83 @@ pub fn completion(cx: &mut Context) {
     );
 }
 
+fn completion_sources(cx: &mut Context) {
+    use helix_lsp::lsp;
+    let (view, doc) = current_ref!(cx.editor);
+
+    let text = doc.text().slice(..);
+    let cursor = doc.selection(view.id).primary().cursor(text);
+
+    let trigger_offset = cursor;
+
+    // TODO: trigger_offset should be the cursor offset but we also need a starting offset from where we want to apply
+    // completion filtering. For example logger.te| should filter the initial suggestion list with "te".
+
+    use helix_core::chars;
+    let mut iter = text.chars_at(cursor);
+    iter.reverse();
+    let offset = iter.take_while(|ch| chars::char_is_word(*ch)).count();
+    let start_offset = cursor.saturating_sub(offset);
+    let prefix = text.slice(start_offset..cursor).to_string();
+
+    let text_str = text.to_string();
+    cx.callback(
+        async move {
+            let candidates: std::collections::HashSet<_> =
+                text_str.split(|c: char| !c.is_alphanumeric()).collect();
+            helix_lsp::Result::Ok(serde_json::Value::Array(
+                candidates
+                    .into_iter()
+                    .map(|r| serde_json::json!({"label": r.to_string()}))
+                    .collect(),
+            ))
+        },
+        move |editor, compositor, response: Option<lsp::CompletionResponse>| {
+            let doc = doc!(editor);
+            if doc.mode() != Mode::Insert {
+                // we're not in insert mode anymore
+                return;
+            }
+
+            let mut items = match response {
+                Some(lsp::CompletionResponse::Array(items)) => items,
+                // TODO: do something with is_incomplete
+                Some(lsp::CompletionResponse::List(lsp::CompletionList {
+                    is_incomplete: _is_incomplete,
+                    items,
+                })) => items,
+                None => Vec::new(),
+            };
```

Comment:
This is serializing to json, then immediately unserializing. Use `jobs.callback` directly and avoid the extra future/parsing?

https://github.com/helix-editor/helix/blob/c4e700cf9e539038239d62c2748cc48e4a1d93db/helix-term/src/commands.rs#L106-L112=